lib/SILOptimizer/Utils/LexicalDestroyFolding.cpp

```diff
@@ -0,0 +1,772 @@
+}
+
+bool LexicalDestroyFolding::filterCandidates(SmallVectorImpl<Candidate> &candidates) const {
+  bool anyViable = false;
+
+  // We have some end_borrows that might be candidates for folding.
+  for (unsigned index = 0, count = candidates.size(); index < count; ++index) {
+    auto &candidate = candidates[index];
+    SmallVector<int, 2> rewritableArgumentIndices;
+    auto viability = isViableInstructionSequence(candidate.match, candidate.argumentIndices);
+    switch (viability) {
+    case MatchViability::Viable:
+      candidate.viable = true;
+      anyViable = true;
+      break;
+    case MatchViability::Nonviable:
+      break;
+    case MatchViability::Illegal:
+      return false;
+      break;
+    }
+  }
+
+  return anyViable;
+}
+
+bool LexicalDestroyFolding::findDominatedUsesOfOriginalBorrowee() {
+  auto recordUse = [&](Operand *use) {
+    if (!dominanceTree.dominates(introducer, use->getUser()))
+      return;
+    originalBorroweeUses.push_back(use);
+    originalBorroweeUsers.insert(use->getUser());
+  };
+  for (auto *use : originalBorrowee->getUses()) {
+    auto *user = use->getUser();
+    if (user == introducer)
+      continue;
+    if (use->getOperandOwnership() == OperandOwnership::Borrow) {
+      if (!BorrowingOperand(use).visitScopeEndingUses([&](Operand *end) {
+            if (end->getOperandOwnership() == OperandOwnership::Reborrow) {
+              return false;
+            }
+            recordUse(end);
+            return true;
+          })) {
+        return false;
+      }
+    }
+    recordUse(use);
+  }
+  return true;
+}
+
+bool LexicalDestroyFolding::originalBorroweeHasUsesWithinBorrowScope() {
+  PrunedLiveness liveness;
+  borrowedValue.computeLiveness(liveness);
+  DeadEndBlocks deadEndBlocks(function);
+  return !liveness.areUsesOutsideBoundary(originalBorroweeUses, &deadEndBlocks);
+}
+
+//===----------------------------------------------------------------------===//
+//                             MARK: Predicates
+//===----------------------------------------------------------------------===//
+
+size_t LexicalDestroyFolding::extendedSimpleUseCount(ApplyInst *ai) const {
+  return extendedSimpleUseCountPerApplySite.lookup(ai);
+}
+
+Optional<LexicalDestroyFolding::Match>
+LexicalDestroyFolding::definesMatchingInstructionSequence(
+    SILInstruction *inst) const {
+  // Look specifically for
+  //
+  //   apply
+  //   end_borrow // inst
+  //   ...
+  //   destroy_value %originalBorrowee
+  auto *ebi = dyn_cast<EndBorrowInst>(inst);
+  if (!ebi)
+    return None;
+  auto *dvi =
+      mayBeginOriginalBorroweeDestroyingSequence(ebi->getNextInstruction());
+  if (!dvi)
+    return None;
+  auto *ai = dyn_cast_or_null<ApplyInst>(ebi->getPreviousInstruction());
+  if (!ai)
+    return {{nullptr, ebi, dvi}};
+
+  return {{ai, ebi, dvi}};
+}
+
+DestroyValueInst *
+LexicalDestroyFolding::mayBeginOriginalBorroweeDestroyingSequence(
+    SILInstruction *inst) const {
+  while (inst) {
```

Comment:
nit: I'm not sure why the loop iteration isn't condensed/canonical:
`for (auto *inst = dvi; inst; inst->geNextInstruction())`

Reply:
Thanks.  Fixed.